0036-spsp-pull-payments/0036-spsp-pull-payments.md

```diff
@@ -0,0 +1,190 @@
+    "assetCode": "USD",
+    "assetScale": "2",
+    "secret": "6jR5iNIVRvqeasJeCty6C+YB5X9FhSOUPCL/5nha5Vs="
+}
+```
+
+#### Response
+
+``` http
+HTTP/1.1 200 OK
+Connection: keep-alive
+Content-Length: 36
+Content-Type: application/json; charset=utf-8
+Date: Mon, 04 Feb 2019 23:42:14 GMT
+Server: nginx/1.10.1
+
+{
+    "token": "$alice.com/pullagreement456"
+}
+``` 
+This token is passed to the client that can use it to conduct the pull payment. The parameters are defined as follows:
+
+| Parameter | Type | Description |
+|---|---|---|
+| `amount` | Integer String | `balance.maximum` SHOULD be set to this amount. |
+| `start` | String | _(OPTIONAL)_ `timeline_info.refill_time` SHOULD be set to this time, unless `start` is in the past, in which case `timeline_info.refill_time` SHOULD be set to `now`. This also holds if `start` is not provided. | |
+| `frequency` | String | `frequency_info.type` SHOULD be set to this value.|
+| `interval` | Integer String | `frequency_info.interval` SHOULD be set to this value.|
+| `cycles` | Integer String | Is used to calculate `timeline_info.expiry_time`, which SHOULD equal `timeline_info.refill_time` + `cycles` * `frequency_info.interval \|\| frequency_info.type`, e.g. `"2019-07-10T01:01:12Z"` = `"2019-02-10T01:01:13Z"` + `5` * `1 MONTH` |
+| `assetCode`| String | `asset_info.code` SHOULD be set to this value. |
+| `assetScale` | Integer String | `asset_info.scale` SHOULD be set to this value. |
+| `secret` | 32 bytes, [base64 encoded](https://en.wikipedia.org/wiki/Base64) (including padding) | Secret that MUST be provided by the client in order to make the pull payment. |
+
+Prior to the creation of the token, the client and server applications MAY negotiate the above defined parameters. This negotiation process is part of the end-user application. 
+
+All requests MUST be exchanged over HTTPS.
+
+### Conducting the pull payment
+
+We assume that the client knows the server's SPSP endpoint (see [Payment Pointers](../0026-payment-pointers/0026-payment-pointers.md)).
+
+1. The user's SPSP client queries the server's SPSP Endpoint.
+
+2. The SPSP endpoint responds with the server info, including the server's ILP address (`destination_account`) and the shared secret (`shared_secret`) to be used in STREAM. 
+    * The `destination_account` SHOULD be used as the STREAM destinationAccount.
+    * The `shared_secret` SHOULD be decoded from base64 and used as the STREAM sharedSecret.
+  
+    It MAY respond with additional information if it is an invoice server (see [SPSP Push Payments](../0035-spsp-push-payments/0035-spsp-push-payments.md)) or a pull payment server. If the amounts within `balance` are negative, the client can push to this endpoint. 
+
+3. The SPSP client establishes a STREAM connection to the server using the server's ILP address and shared secret and constructs a stream to the client on the ILP/STREAM connection.
+
+4. The SPSP server begins sending ILP packets to fulfill the payment.
```

Comment:
Some mention of how the server identifies that this connection is related to the SPSP request would be useful.

Is the ILP address unique per transaction or is the identifier the secret? 

When a connection is established with a specific ILP address and secret I assume the server knows which payment it relates to and so can set the appropriate `sendMax`?

Is there a way to do this that doesn't require the server to store this data somewhere (i.e. allow the server to be stateless)?

Reply:
Instead of storing the information, the token could be a JWT. The client could connect to the endpoint without suffix, i.e. to `/.well-known/pay` instead of `/tokenUUID` and if the first thing that is received by the server is a JWT, it parses and validates it and streams the according funds. 
However, one would probably still have to store balances somewhere to make sure that the client doesn't connect and pull more than allowed.